cf-testd/cf-testd.c

```diff
@@ -230,7 +220,54 @@ bool CFTestD_GetServerQuery(
         return false;
     }
 
-    return CFTestD_ReturnQueryData(conn, query);
+    if (config->report_len == 0)
+    {
+        Log(LOG_LEVEL_INFO,
+            "No report file argument so returning canned data from enterprise plugin server\n");
+        return CFTestD_ReturnQueryData(conn, query);
+    }
+
+    Log(LOG_LEVEL_INFO,
+        "Report file argument specified. Returning report of length %d",
+        config->report_len);
+
+    char *header;
+
+    // TODO for variables reports we will eventually need to split on \m
+    // instead of \n since variable values can contain '\n's.
+
+    size_t num_items =
+        StringCountTokens(config->report, config->report_len, "\n");
+
+    if (num_items < 3)
+    {
+        Log(LOG_LEVEL_ERR,
+            "report file must have a CFR timestamp header line and at least two lines of items\n");
+        return false;
+    }
+
+    StringRef ts_ref =
+        StringGetToken(config->report, config->report_len, 0, "\n");
+    char *ts = xstrndup(ts_ref.data, ts_ref.len);
+    header   = StringFormat("CFR: 0 %s %d\n", ts, config->report_len);
+    SendTransaction(
+        conn->conn_info, header, SafeStringLength(header), CF_MORE);
+
+    for (ssize_t i = 1; i < num_items; i++)
+    {
+        StringRef item_ref =
+            StringGetToken(config->report, config->report_len, i, "\n");
+        char *item = xstrndup(item_ref.data, item_ref.len);
+
+        SendTransaction(
+            conn->conn_info, item, SafeStringLength(item), CF_MORE);
+    }
+
+    const char end_reply[] = "QUERY complete";
+    SendTransaction(
+        conn->conn_info, end_reply, SafeStringLength(end_reply), CF_DONE);
+
```

Comment:
This function has quite a few repeated derefs. I doubt it makes a performance difference, but I think it makes the code more readable if you deref once. I.e. put something like this further up in the function:
```